core/src/main/java/org/apache/accumulo/core/iterators/ClientIteratorEnvironment.java

```diff
@@ -0,0 +1,207 @@
+
+  private final Optional<IteratorScope> scope;
+  private final boolean isFullMajorCompaction;
+  private final Optional<Authorizations> auths;
+  private final boolean isUserCompaction;
+  private final Optional<TableId> tableId;
+  private final Optional<SamplerConfiguration> samplerConfig;
+  private final Optional<ClientServiceEnvironmentImpl> env;
+
+  private ClientIteratorEnvironment(Builder builder) {
+    this.scope = builder.scope;
+    this.isFullMajorCompaction = builder.isFullMajorCompaction;
+    this.auths = builder.auths;
+    this.isUserCompaction = builder.isUserCompaction;
+    this.tableId = builder.tableId;
+    this.samplerConfig = builder.samplerConfig;
+    this.env = builder.env;
+  }
+
+  private ClientIteratorEnvironment(ClientIteratorEnvironment copy) {
+    this.scope = copy.scope.isPresent() ? Optional.of(copy.scope.orElseThrow()) : Optional.empty();
+    this.isFullMajorCompaction = copy.isFullMajorCompaction;
+    this.auths = copy.auths.isPresent() ? Optional.of(copy.auths.orElseThrow()) : Optional.empty();
+    this.isUserCompaction = copy.isUserCompaction;
+    this.tableId =
+        copy.tableId.isPresent() ? Optional.of(copy.tableId.orElseThrow()) : Optional.empty();
+    this.samplerConfig = copy.samplerConfig.isPresent()
+        ? Optional.of(copy.samplerConfig.orElseThrow()) : Optional.empty();
+    this.env = copy.env.isPresent() ? Optional.of(copy.env.orElseThrow()) : Optional.empty();
+  }
+
+  @Override
+  @Deprecated(since = "2.0.0")
+  public SortedKeyValueIterator<Key,Value> reserveMapFileReader(String mapFileName)
+      throws IOException {
+    Configuration hadoopConf = new Configuration();
+    FileSystem fs = FileSystem.get(hadoopConf);
+    return new MapFileIterator(fs, mapFileName, hadoopConf);
+  }
+
+  @Override
+  public IteratorScope getIteratorScope() {
+    return scope.orElseThrow();
+  }
+
+  @Override
+  public boolean isFullMajorCompaction() {
+    if (getIteratorScope() != IteratorScope.majc) {
+      throw new IllegalStateException("Iterator scope is not majc");
+    }
+    return isFullMajorCompaction;
+  }
+
+  @Override
+  @Deprecated(since = "2.0.0")
+  public void registerSideChannel(SortedKeyValueIterator<Key,Value> iter) {
+    throw UOE;
+  }
+
+  @Override
+  public Authorizations getAuthorizations() {
+    if (getIteratorScope() != IteratorScope.scan) {
+      throw new IllegalStateException("Iterator scope is not majc");
+    }
+    return auths.orElseThrow();
+  }
+
+  @Override
+  public IteratorEnvironment cloneWithSamplingEnabled() {
+    if (!isSamplingEnabled()) {
+      throw new SampleNotPresentException();
+    }
+    return new ClientIteratorEnvironment(this);
+  }
+
+  @Override
+  public boolean isSamplingEnabled() {
+    return this.samplerConfig.isPresent();
+  }
+
+  @Override
+  public SamplerConfiguration getSamplerConfiguration() {
+    return samplerConfig.orElse(null);
+  }
```

Comment:
Sampling being enabled and the config being present are separate.  An iterator w/o sampling can create deep copy over sample data.   For this to work, can have an iter env that has sample config but does not have sampling enabled.  To create the deep copy cloneWithSamplingEnabled() is called which returns a new iter env w/ the sampling config and sampling enabled.

So would need a `boolean isSamplingEnabled` instance variable on this class.  

```suggestion
  public IteratorEnvironment cloneWithSamplingEnabled() {
    if (!samplerConfig.isPresent()) {
      throw new SampleNotPresentException();
    }
    // return a new iter env that is identical in every way except it has sampling enabled.
    return new ClientIteratorEnvironment(this, true);
  }

  @Override
  public boolean isSamplingEnabled() {
    return isSamplingEnabled;
  }

  @Override
  public SamplerConfiguration getSamplerConfiguration() {
    if(!isSamplingEnabled){
       // javadoc mentions null should be returned when sampling is not enabled
       return null;
    }
    return samplerConfig.orElse(null);
  }
```

Reply:
Fixed in 3cec5f7